### DISCUSSION TITLE
Move to esp-idf framework

### DISCUSSION THREAD

**Discussion by nliaudat** [nliaudat/floor-heating-controller#5] [Announcements]
The async_tcp is lib is now available in esp-idf (as well as webserver)

With esp-idf, we have better native support of ADC, OTA and memory management

ESP32-C5 and ESP32-C6 are now supported

- Native support for FreeRTOS
- Better support for multi-core CPUs
- Support for more advanced hardware features, such as Bluetooth, Wi-Fi, and low-power modes

**Comment by nstadigs** [comment]
Seems like a good idea. This would also open up to having custom code that handles all the valves in the same loop that would fix some scheduling stuff like the current "valve maintenance" logic that has race conditions. It's hard to fix in the current esphome setup. With custom code we could implement a proper state machine and maintain a queue so that no valve ever falls behind. Hitting two birds with one stone. I've been experimenting with a custom component for esphome to solve this, but it doesn't integrate very cleanly. Doing that for esp-idf would make more sense, I think.

**Reply by nstadigs** [reply]
Oh sorry, I thought you where meaning to move to a custom firmware based on esp-idf. Ignore the previous comment :blush:

**Reply by nliaudat** [reply]
You can make a custom component for esphome if you want ;)
